File: engine-core/src/addressing.rs
//! Deterministic Component addressing.
//!
//! Solana stores state in accounts at Program-Derived Addresses (PDAs), where a
//! PDA is `find_program_address(seeds, program_id)`. Loom fixes the seed scheme so
//! every Component account in every world is addressable from
//! `(world_id, entity_id, component_id)` alone, with no external index:
//!
//! ```text
//! seeds = [ b"loom", b"cmp", world_id.le, entity_id.le, component_id.le ]
//! ```
//!
//! The on-chain program passes these seeds to `find_program_address` for the real
//! account address. Off-chain (indexer, SDK, this core) we want the same identity
//! without ed25519 curve math, so we derive a 32-byte **logical address** by
//! hashing the seeds with FNV-1a, domain-separated per 8-byte lane. The Rust core
//! and TypeScript SDK compute it identically; it is the join key the indexer uses
//! to reconstruct world state from raw Component accounts.

use crate::hash::Hasher;
use crate::ids::{ComponentId, EntityId, WorldId};

/// Domain tag mixed into every Component address derivation.
pub const COMPONENT_TAG: &[u8] = b"loom:cmp";

/// A 32-byte logical address identifying one Component record.
pub type Address = [u8; 32];

/// Derive the logical address of the Component record at
/// `(world_id, entity_id, component_id)`.
///
/// Mirrored exactly by `deriveComponentAddress` in the TypeScript SDK.
pub fn component_address(
    world: WorldId,
    entity: EntityId,
    component: ComponentId,
) -> Address {
    let mut out = [0u8; 32];
    // Four independent 8-byte lanes, each domain-separated by its lane index, so
    // the full 32 bytes carry 256 bits of derived identity rather than a repeated
    // 64-bit value.
    for (lane, chunk) in out.chunks_mut(8).enumerate() {
        let mut h = Hasher::new();
        h.write(COMPONENT_TAG)
            .write_u8(lane as u8)
            .write_u64(world)
            .write_u64(entity)
            .write_u32(component);
        chunk.copy_from_slice(&h.finish().to_le_bytes());
    }
    out
}

/// Lowercase hex rendering of an [`Address`], for logs and the indexer's keys.
pub fn to_hex(addr: &Address) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut s = String::with_capacity(64);
    for &b in addr {
        s.push(HEX[(b >> 4) as usize] as char);
        s.push(HEX[(b & 0x0f) as usize] as char);
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deterministic_and_distinct() {
        let a = component_address(1, 1, 1);
        assert_eq!(a, component_address(1, 1, 1), "must be deterministic");
        assert_ne!(a, component_address(2, 1, 1), "world distinguishes");
        assert_ne!(a, component_address(1, 2, 1), "entity distinguishes");
        assert_ne!(a, component_address(1, 1, 2), "component distinguishes");
    }

    #[test]
    fn lanes_are_independent() {
        // If lane domain separation were missing, all 4 lanes would be equal.
        let a = component_address(7, 42, 3);
        let lane0 = &a[0..8];
        let lane1 = &a[8..16];
        assert_ne!(lane0, lane1);
    }

    #[test]
    fn hex_is_64_chars() {
        let a = component_address(1, 1, 1);
        assert_eq!(to_hex(&a).len(), 64);
    }
}

File: sdk/src/addressing.ts
// Deterministic Component addressing. `componentAddress` is the join key the
// indexer uses to reconstruct world state from raw Component accounts, derived
// from the PDA seed scheme:
//
//   seeds = [ "loom", "cmp", world_id.le, entity_id.le, component_id.le ]

import { Hasher } from "./hash.ts";

const COMPONENT_TAG = new TextEncoder().encode("loom:cmp");

/** A 32-byte logical Component address. */
export type Address = Uint8Array;

/** Derive the logical address of the Component at (world, entity, component). */
export function componentAddress(
  world: bigint,
  entity: bigint,
  component: number,
): Address {
  const out = new Uint8Array(32);
  const dv = new DataView(out.buffer);
  for (let lane = 0; lane < 4; lane++) {
    const word = new Hasher()
      .write(COMPONENT_TAG)
      .writeU8(lane)
      .writeU64(world)
      .writeU64(entity)
      .writeU32(component)
      .finish();
    dv.setBigUint64(lane * 8, word, true); // little-endian
  }
  return out;
}

/** Lowercase hex of an address (64 chars). */
export function toHex(addr: Address): string {
  let s = "";
  for (const b of addr) s += b.toString(16).padStart(2, "0");
  return s;
}

/**
 * Parse a 64-char lowercase hex string back into an address - the inverse of
 * `toHex`. Returns undefined unless the input is exactly 32 bytes of hex.
 */
export function fromHex(s: string): Address | undefined {
  if (!/^[0-9a-f]{64}$/.test(s)) return undefined;
  const out = new Uint8Array(32);
  for (let i = 0; i < 32; i++) out[i] = Number.parseInt(s.slice(2 * i, 2 * i + 2), 16);
  return out;
}
